{% extends "layout.html" %}
{% block body %}
    <div class="container" style="font-size: 16px;">
        <h1>Frequently Asked Questions</h1>
        <br/>
        <ul class="media-list">
            <li class="media">
                <div class="media-body">
                    <h4 class="media-heading">I have identified a rare variant in ExAC that I believe is associated with a specific clinical phenotype. What phenotype data are available for these individuals?</h4>
                    Many of the individuals who have contributed data to ExAC were not fully consented for phenotype data sharing, and unfortunately at this time we are typically unable to provide any information about the clinical status of variant carriers. We have made every effort to exclude individuals with severe pediatric diseases from the ExAC data set, and certainly do not expect our data set to be <i>enriched</i> for such individuals, but we typically cannot rule out the possibility that some of our participants do actually suffer from your disease of interest.
                    <br><br>
                    In the future it may be possible to obtain phenotype data for a subset of ExAC samples. If you are interested in using such data in your studies, please <a href="mailto:exomeconsortium@gmail.com">email us</a>.
                </div>
            </li>
            <li class="media">
                <div class="media-body">
                    <h4 class="media-heading">What genome build is the ExAC data based on?</h4>
                    All data are based on <a href="ftp://ftp.broadinstitute.org/pub/seq/references/Homo_sapiens_assembly19.fasta">GRCh37/hg19</a>
                </div>
            </li>
            <li class="media">
                <div class="media-body">
                    <h4 class="media-heading">What version of dbSNP was used to annotate variants?</h4>
                    dbSNP 135. The browser supports RSIDs up to dbSNP 141, but these are not included in the VCF.
                </div>
            </li>
            <li class="media">
                <div class="media-body">
                    <h4 class="media-heading">What version of Gencode was used to annotate variants?</h4>
                    Version 19 (annotated with VEP version 77).
                </div>
            </li>
            <li class="media">
                <div class="media-body">
                    <h4 class="media-heading">What are the restrictions on data usage?</h4>
                    All data here are released under a <a href="http://www.wellcome.ac.uk/stellent/groups/corporatesite/@policy_communications/documents/web_document/wtd003207.pdf">Fort Lauderdale Agreement</a> for the benefit of the wider biomedical community. You can freely download and search the data, and use it for publications focused on specific sets of variants (for instance, assessing the frequency of a set of candidate causal variants observed in a collection of rare disease patients). However, we ask that you not publish global (genome-wide) analyses of these data until after the ExAC flagship paper has been published, estimated to be in early 2015. If youâ€™re uncertain which category your analyses fall into, please <a href="mailto:exomeconsortium@gmail.com">email us</a>.
                </div>
            </li>
            <li class="media">
                <div class="media-body">
                    <h4 class="media-heading">Are all the individuals in the <a href="http://evs.gs.washington.edu/EVS/">Exome Variant Server</a> included?</h4>
                    No. We were not given permission from dbGaP to include individuals from several of the cohorts included in the NHLBI's Exome Sequencing Project. As a result, genuine rare variants that are present in the EVS may not be observed in ExAC.
                </div>
            </li>
            <li class="media">
                <div class="media-body">
                    <h4 class="media-heading">How should I cite discoveries made using ExAC data?</h4>
                    We anticipate the publication of a flagship paper describing this data set in early 2015. Until that is published, please cite: Exome Aggregation Consortium (ExAC), Cambridge, MA (URL: http://exac.broadinstitute.org) [date (month, year) accessed].
                </div>
            </li>
            
            
        </ul>
    </div>
{% endblock %}
