{% extends "layout.html" %}
{% block body %}
    <!-- Render context vars in JS here -->
    <script type="text/javascript">
        window.gene = {{ gene|tojson|safe }};
        window.transcript = {{ transcript|tojson|safe }};
        window.variants_in_gene = {{ variants_in_gene|tojson|safe }};
        window.variants_in_transcript = {{ variants_in_transcript|tojson|safe }};
        window.lof_variants_in_gene = {{ lof_variants_in_gene|tojson|safe }};
        window.transcripts_in_gene = {{ transcripts_in_gene|tojson|safe }};
        window.coverage_stats = {{ coverage_stats|tojson|safe }};

    </script>
    <style>

    </style>
    <script>
        $(document).ready(function() {
            $("#variants_loading").hide();
            $("#variants_table_container").show();
            $("#gene_variants_table").tablesorter({widthFixed: true})
                    .tablesorterPager({container: $("#pager")});
        });
    </script>
    <div class="container">
        <h1>Gene: {{ gene.gene_name }}</h1>
        <hr/>
        <div class="row">
            <div class="col-md-6">
                <dl class="dl-horizontal">
                    <dt>Number of variants</dt>
                    <dd><span id="number_of_variants"></span> (Including filtered: {{ variants_in_gene|length }})</dd>
                    <dt>Number of LoF variants</dt>
                    <dd><span id="number_of_lof_variants"></span> (Including filtered: {{ lof_variants_in_gene|length }})</dd>
                    <script>
                        filtered_variants = _.filter(window.variants_in_gene, function(d) {
                            return d.filter == 'PASS';
                        });
                        $('#number_of_variants').replaceWith(filtered_variants.length);
                        lof_filtered_variants = _.filter(window.variants_in_gene, function(d) {
                            return d.filter == 'PASS' && d.category == 'lof_variant';
                        });
                        $('#number_of_lof_variants').replaceWith(lof_filtered_variants.length);
                    </script>
                    <dt>LoF rate</dt>
                    <dd>{{ '%0.4g' % composite_lof_frequency }}</dd>
                    <dt>GeneCards</dt>
                    <dd>
                        <a href="http://www.genecards.org/cgi-bin/carddisp.pl?gene={{ gene.gene_name }}" target="_blank">
                            {{ gene.gene_name }}
                            <i class="fa fa-external-link"></i>
                        </a>
                    </dd>
                    <dt>Ensembl</dt>
                    <dd>
                        <a href="http://grch37.ensembl.org/Homo_sapiens/Gene/Summary?g={{ gene.gene_id }}">{{ gene.gene_id }}</a>
                        <i class="fa fa-external-link"></i>
                    </dd>
                    <dt>UCSC</dt>
                    <dd>
                        <a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&position=chr{{ gene.chrom }}%3A{{ gene.start - 1 }}-{{ gene.stop - 1 }}" target="_blank">
                            {{ gene.chrom }}:{{ gene.start - 1 }}-{{ gene.stop - 1 }}
                            <i class="fa fa-external-link"></i>
                        </a>
                    </dd>
                    {% if gene.omim_accession %}
                        <dt>OMIM</dt>
                        <dd>
                            <a href="http://omim.org/entry/{{ gene.omim_accession }}" target="_blank">
                                {{ gene.gene_name }}
                                <i class="fa fa-external-link"></i>
                            </a>
                        </dd>
                    {% endif %}
                    <dt>GTEx (Expression)</dt>
                    <dd>
                        <a href="http://www.gtexportal.org/home/gene/{{ gene.gene_name }}" target="_blank">
                            {{ gene.gene_name }}
                            <i class="fa fa-external-link"></i>
                        </a>
                    </dd>
                </dl>
            </div>
            <div class="col-md-6">
                <dl class="dl-horizontal">
                    <dt>Position</dt>
                    <dd>{{ gene.chrom }}:{{ gene.start }}-{{ gene.stop }}</dd>
                </dl>
                <div class="dropdown">
                    <button class="btn btn-default dropdown-toggle" type="button" id="transcript_dropdown" data-toggle="dropdown">
                        Transcripts
                        <span class="caret"></span>
                    </button>
                    <ul class="dropdown-menu" role="menu" aria-labelledby="transcript_dropdown">
                        {% for transcript in transcripts_in_gene %}
                            <li role="presentation">
                                <a role="menuitem" tabindex="-1" href="/transcript/{{ transcript.transcript_id }}">
                                    {{ transcript.transcript_id }}
                                    {% if transcript.transcript_id == gene.canonical_transcript %}
                                        *
                                    {% endif %}
                                </a>
                            </li>
                        {% endfor %}
                    </ul>
                </div>
            </div>
        </div>
        <div class="row">
            <h3>Gene summary</h3>
            <h4>(Coverage shown for canonical transcript: {{ transcript.transcript_id }})</h4>
            <div class="row">
                {% include 'transcript_display.html' %}
            </div>
            <div class="row">
                {% if variants_in_gene %}
                    {% with %}
                        {% set table_variants = variants_in_gene %}
                        {% include 'variant_table.html' %}
                    {% endwith %}
                {% else %}
                    No variants in this gene.
                {% endif %}
            </div>
        </div>
    </div>
{% endblock %}