{% include 'coverage_selectors.html' %}
<br/>
<div id="gene_plot_container"></div>
{% include 'variant_selectors.html' %}
<script>

// this should be global - should not vary across installation
var EXON_PADDING = 20;

// todo: move this somewhere else
/*
    The following methods are for working with "Coding Coordinates",
    a coordinate space that we use to plot data in the coding regions of a transcript.

    Conceptually: suppose you lined up all the coding regions in a transcript, with some padding on each side,
    then plotted any variants that overlap. The position of a variant on this line is the coding position -
    this will obviously differ from the actual genomic or cds coordinates.

    Random notes:
        - coding coordinates have no concept of a gene - they are solely a property of a transcript
        - should probably have a map between coding coodinates and protein position
        - Brett, will you please write some fucking tests for this...

 */
window.get_coding_coordinates = function(_transcript, position_list) {
    var num_exons = _transcript.exons.length;
    var exon_offsets = [];
    // initialize with one sided padding
    for (var i=0; i<num_exons; i++) {
        exon_offsets.push(EXON_PADDING);
    }
    for (var i=0; i<num_exons; i++) {
        for (var j=i+1; j<num_exons; j++) {
            exon_offsets[j] += _transcript.exons[i]['stop'] - _transcript.exons[i]['start'] + EXON_PADDING;
        }
    }

    // get each position
    // todo: optimize by sorting positions
    var coding_positions = [];
    for (var i=0; i<num_exons; i++) {  // todo: underscore init method?
        coding_positions.push(-1);
    }
    _.each(position_list, function(position, i) {
        _.each(transcript.exons, function(exon, j) {
            if (position >= exon.start && position <= exon.stop) {
                coding_positions[i] = exon_offsets[j] + position - exon.start;
                return;
            }
        });
    });
    return coding_positions;
};

window.get_coding_coordinate = function(_transcript, position) {
    return get_coding_coordinates(_transcript, [position])[0];
};


window.get_coding_coordinate_params = function(_transcript) {
    var ret = {};
    ret.num_exons = _transcript.exons.length;
    ret.size = EXON_PADDING;
    _.each(_transcript.exons, function(exon) {
        ret.size += exon.stop - exon.start + EXON_PADDING;
    });
    return ret;
};

// Gene charts
gene_chart_margin = {top: 10, right: 30, bottom: 30, left: 80},
    gene_chart_margin_lower = {top: 5, right: gene_chart_margin.right, bottom: 5, left: gene_chart_margin.left},
    gene_chart_width = 1100 - gene_chart_margin.left - gene_chart_margin.right;

lower_gene_chart_height = 50 - gene_chart_margin_lower.top - gene_chart_margin_lower.bottom,
    gene_chart_height = 300 - gene_chart_margin.top - gene_chart_margin.bottom - lower_gene_chart_height - gene_chart_margin_lower.top - gene_chart_margin_lower.bottom;

function gene_chart(data, exon_data, variant_data) {
    var metric = 'mean';
    var transcript = exon_data[0].transcript_id;
    var padding = 20;
    var total_exon_length = 0;
    var total_exon_length_padded = 0;
    var running_exon_length = [];
    $.each(exon_data, function(i, x) {
        total_exon_length += (x.stop - x.start + 1);
        total_exon_length_padded += (x.stop - x.start + 1 + padding);
        running_exon_length.push(total_exon_length);
    });
//    console.log("Total length: ", total_exon_length);
//    console.log("Total length padded: ", total_exon_length_padded);
//    console.log("Running lengths: ", running_exon_length);
    console.log('total exon length: ', total_exon_length_padded);
    console.log('width: ', gene_chart_width);

    var start_pos = exon_data[0].start;
    var exon_x_scale = d3.scale.linear()
        .domain([0, total_exon_length_padded - padding])
        .range([0, gene_chart_width]);

    var y = d3.scale.linear()
        .domain([0, d3.max(data, function(d) { return d[metric]; })])
        .range([gene_chart_height, 0]);

    var xAxis = d3.svg.axis()
        .scale(exon_x_scale)
        .orient("bottom");

    var yAxis = d3.svg.axis()
        .scale(y)
        .orient("left");

    var svg = d3.select('#gene_plot_container').append("svg")
        .attr("width", gene_chart_width + gene_chart_margin.left + gene_chart_margin.right)
        .attr("height", gene_chart_height + gene_chart_margin.top + gene_chart_margin.bottom)
        .append("g")
        .attr('id', 'inner_graph')
        .attr("transform", "translate(" + gene_chart_margin.left + "," + gene_chart_margin.top + ")");

    var new_data = _.filter(data, function(x) { return x.exon_number >= 0; });

    console.log('Exon data: ', new_data);
    svg.selectAll("bar")
        .data(new_data)
        .enter()
        .append("rect")
        .attr('class', 'main_plot_bars')
        .style("fill", "steelblue")
        .attr("x", function(d, i) {
//            console.log("i: ", i);
//            console.log("d: ", d);
            var relative_start_pos = i;
            if (d.exon_number > 0) {
                relative_start_pos += (d.exon_number)*padding;
            }
            return exon_x_scale(relative_start_pos);
        })
        .attr("width", 1)
        .attr("y", function(d) { return y(d[metric]); })
        .attr("height", function(d) { return gene_chart_height - y(d[metric]); });

    svg.append("g")
        .attr("class", "x axis")
        .attr("transform", "translate(0," + gene_chart_height + ")")
        .call(xAxis);

    svg.append("g")
        .attr("class", "y axis")
        .call(yAxis);

    var svg_outer = d3.select('#gene_plot_container').append("svg")
        .attr("width", gene_chart_width + gene_chart_margin_lower.left + gene_chart_margin_lower.right)
        .attr("height", lower_gene_chart_height)
        .append("g")
        .attr('id', 'track')
        .attr("transform", "translate(" + gene_chart_margin_lower.left + "," + 0 + ")");

    var exon_color = "lightsteelblue";
    svg_outer.append("line")
        .attr("y1", lower_gene_chart_height/2)
        .attr("y2", lower_gene_chart_height/2)
        .attr("x1", 0)
        .attr("x2", exon_x_scale(data.length))
        .attr("stroke-width", 10)
        .attr("stroke", exon_color);

    svg_outer.selectAll("bar")
        .data(exon_data)
        .enter()
        .append("rect")
        .attr('class', 'track_bar')
        .style("fill", exon_color)
        .attr("x", function(d, i) {
            var relative_start_pos;
            if (i == 0) {
                relative_start_pos = 0;
            } else {
                relative_start_pos = running_exon_length[i-1] + i*padding;
            }
            return exon_x_scale(relative_start_pos);
        })
        .attr("y", 0)
        .attr("rx", 6)
        .attr("ry", 6)
        .attr("width", function(d, i) { return exon_x_scale(d.stop-d.start+1); })
        .attr("height", lower_gene_chart_height);

//    console.log("Variant data", variant_data);

    var variant_size_scale = d3.scale.log()
        .domain([d3.min(variant_data, function(d) { return d.allele_freq; }), d3.max(variant_data, function(d) { return d.allele_freq; })])
        //Circle/Ellipse
        .range([2, lower_gene_chart_height/3]);
        //Rectangle
//        .range([lower_gene_chart_height, 2]);

    var tip = d3.tip().attr('class', 'd3-tip').html(function(d) { return d.category; });
    svg.call(tip);

    svg_outer.selectAll("bar")
        .data(variant_data)
        .enter()
        .append("a")
        .attr('class', 'track_variant_link')
        .attr("xlink:href", function(d, i) { return "/variant/" + d.chrom + "-" + d.pos + "-" + d.ref + "-" + d.alt; })
        .attr("data-toggle", "tooltip")
        .attr('filter_status', function(d) {
            return d.filter;
        })
        .attr('category', function(d) {
            return d.category;
        })
{#        .attr("title", function(d) {#}
{#            return d.vep_annotations[0]['Consequence'];#}
{#        })#}
        .on('mouseover', function(d) {
            $('#variant_' + d.variant_id).find('td').addClass('table_hover');
            tip.show(d);
        })
        .on('mouseout', function(d) {
            $('#variant_' + d.variant_id).find('td').removeClass('table_hover');
            tip.hide(d);
        })
        //Circle
//        .append("circle")
        //Ellipse
        .append("ellipse")
        .attr("class", "track_variant")
        .style("fill", function(d) {
            return get_color(d);
        })
        .style("opacity", 0.5)
        .attr("cx", function(d, i) {
            var tx_coord = d.transcript_coordinates[transcript];
            if (tx_coord == 0) {
                return -1000;
            } else {
                if (d.vep_annotations[0]['EXON'] != '') {
                    var variant_exon_info;
                    variant_exon_info = d.vep_annotations[0]['EXON'].split('/');
                    var variant_exon_number;
                    if (d.vep_annotations[0]['STRAND'] == '-1') {
                        variant_exon_number = variant_exon_info[1] - variant_exon_info[0];
                    } else {
                        variant_exon_number = variant_exon_info[0] - 1;
                    }
{#                    console.log(d);#}
{#                    console.log(tx_coord, " ", variant_exon_number);#}
                    return exon_x_scale(tx_coord + variant_exon_number*padding);
                } else {
                    //TODO: Implement correct intron drawing
                    var variant_intron_info;
                    variant_intron_info = d.vep_annotations[0]['INTRON'].split('/');

                    var variant_intron_number;
                    if (d.vep_annotations[0]['STRAND'] == '-1') {
                        variant_intron_number = variant_intron_info[1] - variant_intron_info[0];
                    } else {
                        variant_intron_number = variant_intron_info[0] - 1;
                    }
                    return 0;
                }
            }
        })
        .attr("cy", lower_gene_chart_height/2)
        //Circle
//        .attr("r", function(d, i) { return variant_size_scale(d.allele_freq); })
        //Ellipse
        .attr("rx", 2)
        .attr("ry", function(d, i) { return variant_size_scale(d.allele_freq); });
        //Rectangle
//        .append("rect")
//        .attr("class", "track_variant")
//        .style("fill", "darkred")
//        .style("opacity", 0.5)
//        .attr("x", function(d, i) {
//            var tx_coord = d.transcript_coordinates[transcript];
//            if (tx_coord == 0) {
//                return -1000;
//            } else {
//                var variant_exon_number = d.vep_annotations[0]['EXON'].split('/')[0] - 1;
//                return exon_x_scale(tx_coord + variant_exon_number*padding);
//            }
//        })
//        .attr("y", function(d, i) { return lower_gene_chart_height/2 - variant_size_scale(d.allele_freq)/2; } )
//        .attr("width", 2)
//        .attr("height", function(d, i) { return variant_size_scale(d.allele_freq); })
//        .attr("rx", 6)
//        .attr("ry", 6);
}

function change_gene_chart_variant_size(variant_data, change_to) {
    var svg_outer = d3.select('#gene_plot_container').select('#track');

    var variant_size_scale;
    if (change_to) {
        variant_size_scale = d3.scale.log()
            .domain([d3.min(variant_data, function (d) {
                return d.allele_freq;
            }), d3.max(variant_data, function (d) {
                return d.allele_freq;
            })])
            .range([lower_gene_chart_height / 3, 2]);
    } else {
        variant_size_scale = d3.scale.log()
            .domain([d3.min(variant_data, function (d) {
                return d.allele_freq;
            }), d3.max(variant_data, function (d) {
                return d.allele_freq;
            })])
            .range([2, lower_gene_chart_height / 3]);
    }
    svg_outer.selectAll("a")
        .selectAll("ellipse")
        .transition()
        .duration(500)
        .attr("ry", function(d, i) { return variant_size_scale(d.allele_freq); });
}

function change_gene_chart_metric(data, metric) {
    var y = d3.scale.linear()
        .domain([0, d3.max(data, function(d) { return d[metric]; })])
        .range([gene_chart_height, 0]);

    var svg = d3.select('#gene_plot_container').select('#inner_graph');

    var new_data = _.filter(data, function(x) { return x.exon_number >= 0; });

//    console.log('Exon data: ', exon_data);
    svg.selectAll("rect")
        .data(new_data)
        .transition()
        .duration(500)
        .attr("y", function(d) { return y(d[metric]); })
        .attr("height", function(d) { return gene_chart_height - y(d[metric]); });

    var yAxis = d3.svg.axis()
        .scale(y)
        .orient("left");

    svg.select(".y.axis")
        .transition()
        .duration(200)
        .call(yAxis);

}

$(document).ready(function() {
// Change coverage plot
    $('.coverage_metric_buttons').change(function () {
        var v = $(this).attr('id').replace('_covmet_button', '');
        $('.coverage_subcat_selectors').hide();
        if (v == 'covered') {
            $('#over_x_select_container').show();
            v = $('#over_x_select').val();
        } else {
            $('#average_select_container').show();
            v = $("#average_select").val();
        }
        change_gene_chart_metric(window.coverage_stats, v);
    });
    $('#over_x_select').change(function () {
        change_gene_chart_metric(window.coverage_stats, $(this).val());
    });
    $('#average_select').change(function () {
        change_gene_chart_metric(window.coverage_stats, $(this).val());
    });

// Change exon diagram
    $('#inverted_checkbox').change(function () {
        setTimeout(function () {
            var v = $('#inverted_checkbox').is(':checked');
            change_gene_chart_variant_size(window.variants_in_transcript, v);
        }, 10);
    });

    $('.consequence_display_buttons, #filtered_checkbox').change(function () {
        setTimeout(update_variants, 10);
    });
});
</script>