{% extends "layout.html" %}
{% block body %}
    <!-- Render context vars in JS here -->
    <script type="text/javascript">
        window.variant = {{ variant|tojson|safe }};
        window.base_coverage = {{ base_coverage|tojson|safe }};
        window.any_covered = {{ any_covered|tojson|safe }};
        window.consequence = {{ consequences|tojson|safe }};
        window.metrics = {{ metrics|tojson|safe }};

        $(document).ready(function() {
{#            if (!window.variant) {#}
                draw_region_coverage(window.base_coverage, 'mean', window.variant.ref);
                $('.coverage_metric_buttons').change(function () {
                    var v = $(this).attr('id').replace('_covmet_button', '');
                    $('.coverage_subcat_selectors').hide();
                    if (v == 'covered') {
                        $('#over_x_select_container').show();
                        v = $('#over_x_select').val();
                    } else {
                        $('#average_select_container').show();
                        v = $("#average_select").val();
                    }
                    draw_region_coverage(window.base_coverage, v, window.variant.ref);
                });
                $('#over_x_select').change(function () {
                    draw_region_coverage(window.base_coverage, $(this).val(), window.variant.ref);
                });
                $('#average_select').change(function () {
                    draw_region_coverage(window.base_coverage, $(this).val(), window.variant.ref);
                });
{#            }#}
        });
    </script>
    <style>
    .d3_graph {
        font: 10px sans-serif;
    }

    .bar rect {
        fill: steelblue;
        shape-rendering: crispEdges;
    }

    .bar text {
        fill: #fff;
    }

    .axis path, .axis line {
        fill: none;
        stroke: #000;
        shape-rendering: crispEdges;
    }
    </style>
    <script>
        var af_buckets = [0.0001, 0.0002, 0.0005, 0.001, 0.002, 0.005, 0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1];
        function get_af_bucket_text(bin) {
            if (bin == 'singleton' || bin == 'doubleton') {
                return 'This is the site quality distribution for all ' + bin + 's in ExAC.';
            } else if (bin == '0.0001') {
                return 'This is the site quality distribution for all variants with AF < ' + bin + ' in ExAC.';
            } else {
                return 'This is the site quality distribution for all variants with ' + af_buckets[af_buckets.indexOf(parseFloat(bin)) - 1] + ' < AF < ' + bin + ' in ExAC.';
            }
        }
        $(document).ready(function() {
            $('.frequency_display_buttons').change(function() {
                $('.frequency_displays').hide();
                var v = $(this).attr('id').replace('_button', '');
                $('#' + v + '_container').show();
            });
            $('#frequency_table').tablesorter({
                sortList: [[4,1], [0,0]]
            });

            if (window.variant != null && 'genotype_depths' in window.variant) {
                draw_quality_histogram(window.variant.genotype_depths[1], '#quality_display_container', false);
                $('.quality_display_buttons').change(function() {
                    var v = $(this).attr('id').replace('_button', '');
                    var f = $('.quality_full_site_buttons.active').attr('id') == 'variant_site_button' ? 0 : 1;
                    draw_quality_histogram(window.variant[v][f], '#quality_display_container', false);
                });
                $('.quality_full_site_buttons').change(function() {
                    var v = $('.quality_display_buttons.active').attr('id').replace('_button', '');
                    var f = $(this).attr('id') == 'variant_site_button' ? 0 : 1;
                    draw_quality_histogram(window.variant[v][f], '#quality_display_container', false);
                });

                // Quality metric histograms
                var data = _.zip(_.map(window.metrics['Site Quality']['mids'], Math.exp), window.metrics['Site Quality']['hist']);
                console.log(data);
                draw_quality_histogram(data, '#quality_metric_container', true);
                var bin = window.metrics['Site Quality']['metric'].split('_')[1];
                $('#site_quality_note').html(get_af_bucket_text(bin));
                var pos = $('#quality_metric_select').val().split(': ')[1];
                add_line_to_quality_histogram(data, pos, '#quality_metric_container', true);
                $('#quality_metric_select').change(function() {
                    var v = $(this).val().split(': ');
                    var log = false;
                    var data;
                    if (v[0] == 'Site Quality') {
                        data = _.zip(_.map(window.metrics[v[0]]['mids'], Math.exp), window.metrics[v[0]]['hist']);
                        log = true;
                        var bin = window.metrics['Site Quality']['metric'].split('_')[1];
                        $('#site_quality_note').html(get_af_bucket_text(bin));
                    } else {
                        data = _.zip(window.metrics[v[0]]['mids'], window.metrics[v[0]]['hist']);
                        $('#site_quality_note').html('');
                    }
                    draw_quality_histogram(data, '#quality_metric_container', log);
                    add_line_to_quality_histogram(data, v[1], '#quality_metric_container', log);
                });
            } else {
                $('#quality_metrics_container').hide();
{#                $('#frequency_info_container').hide();#}
            }
        });
    </script>
    <div class="container">
        <h1><span class="hidden-xs">Variant: </span>{{ variant.chrom }}:{{ variant.pos }} {{ variant.ref }} / {{ variant.alt }}</h1>
        <hr/>

        {% if variant.orig_alt_alleles|length > 1 %}
            <p><span class="label label-info">Note:</span> This variant is multiallelic! The other alt alleles are:</p>
            <ul>
                {% for allele in variant.orig_alt_alleles %}
                    {% if allele != variant.variant_id %}
                        <li>
                            <a href="/variant/{{ allele }}">{{ allele }}</a>
                        </li>
                    {% endif %}
                {% endfor %}
            </ul>
            <hr/>
        {% endif %}

{#          Upper display #}
        {% if variant.variant_id %}
            <div class="row">
                <div class="col-md-6">
                    <dl class="dl-horizontal">
                        <dt>dbSNP</dt>
                        {% if variant.rsid and variant.rsid != "." %}
                            <dd><a href="http://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs={{ variant.rsid }}" target="_blank">{{ variant.rsid }}</a></dd>
                        {% else %}
                            <dd>Not found in dbSNP</dd>
                        {% endif %}
                        <dt>Allele Frequency</dt>
                        <dd>{{ '%0.4g' % variant.allele_freq }}</dd>
                        <dt>Allele Count</dt>
                        <dd>{{ variant.allele_count }} / {{ variant.allele_num }}</dd>
                        <dt>UCSC</dt>
                        <dd>
                            <a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&highlight=hg19.chr{{ variant.chrom }}%3A{{ variant.pos }}-{{ variant.pos + variant.ref|length - 1 }}&position=chr{{ variant.chrom }}%3A{{ variant.pos - 25 }}-{{ variant.pos + variant.ref|length - 1 + 25 }}" target="_blank">
                                {{ variant.variant_id }}
                                <i class="fa fa-external-link"></i>
                            </a>
                        </dd>
                        <dt>ClinVar</dt>
                        <dd>
                            {% if not variant.rsid or variant.rsid == "." %}
                                <a href="http://www.ncbi.nlm.nih.gov/clinvar?term=({{ variant.chrom }}%5BChromosome%5D)%20AND%20{{ variant.pos }}%5BBase%20Position%20for%20Assembly%20GRCh37%5D" target="_blank">
                            {% else %}
                                <a href="http://www.ncbi.nlm.nih.gov/clinvar?term={{ variant.rsid }}%5BVariant%20ID%5D">
                            {% endif %}
                                Click to search for variant in Clinvar
                                <i class="fa fa-external-link"></i>
                            </a>
                        </dd>
                    </dl>
                </div>
                <div class="col-md-6">
                    <dl class="dl-horizontal">
                        <dt>Site Quality</dt>
                        <dd>{{ variant.site_quality }}</dd>
                        <dt>
                            {% if variant.filter != "PASS" %}
                                <span class="label label-danger">Filter Status</span>
                            {% else %}
                                Filter Status
                            {% endif %}
                        </dt>
                        <dd>{{ variant.filter }}</dd>
                    </dl>
                    <div class="row">
                        <div class="col-md-8 col-md-offset-1">
                            <div class="panel-group" id="metrics_outer_accordion" role="tablist" aria-multiselectable="true">
                                <div class="panel panel-default">
                                    <div class="panel-heading" role="tab" id="quality_collapse_heading"   style="height: auto">
                                        <div class="panel-title">
                                            <a data-toggle="collapse" data-parent="#accordion" href="#quality_collapse" aria-expanded="false" aria-controls="quality_collapse">
                                                <small>Additional Quality Metrics</small>
                                            </a>
                                        </div>
                                    </div>
                                    <div id="quality_collapse" class="panel-collapse collapse" role="tabpanel" aria-labelledby="quality_collapse_heading">
                                        <div class="row">
                                            <div class="col-md-10 col-md-offset-1">
                                                <div id="quality_metric_container"></div>
                                                <small><div id="site_quality_note"></div></small>
                                                <small><span class="label label-info">Note:</span> These are site-level quality metrics: they may be unpredictable for multi-allelic sites.</small>
                                                <select id="quality_metric_select" class="form-control">
                                                    <option><small>Site Quality: {{ variant.site_quality }}</small></option>
                                                    {% for metric in variant.quality_metrics %}
                                                        <option><small>{{ metric }}: {{ variant.quality_metrics[metric] }}</small></option>
                                                    {% endfor %}
                                                </select>
                                                <br/>
                                            </div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
            <hr/>
        {% endif %}

{#          Lower display#}
        <div class="row">
            <div class="col-md-6">
                <div id="annotation_container">
                    <div class="section_header">Annotations</div>
                    {% if variant.vep_annotations %}
                        <p>This variant falls on {{ variant.transcripts|length }} transcripts in {{ variant.genes|length }} genes:</p>
                        <div class="panel-group" id="annotation_accordion">
                            {% for consequence in ordered_csqs %}
                                <h4>{{ consequence|replace('_variant', '')|replace('_', ' ')|replace('utr', 'UTR')|replace('3 prime', "3'")|replace('5 prime', "5'")|replace('nc ', "non-coding ") }}</h4>
                                <ul>
                                    {% for gene in consequences[consequence] %}
                                        <li>
                                             <a href="/gene/{{ gene }}">
                                            {% if consequences[consequence][gene][0].SYMBOL %}
                                                {{ consequences[consequence][gene][0].SYMBOL }}
                                            {% else %}
                                                {{ gene }}
                                            {% endif %}
                                            </a>
                                            {% if consequences[consequence][gene]|length > 1 %}
                                                <span class="dropdown">
                                                    <button class="btn btn-default dropdown-toggle" type="button" id="transcript_dropdown" data-toggle="dropdown">
                                                        Transcripts
                                                        <span class="caret"></span>
                                                    </button>
                                                    <ul class="dropdown-menu" role="menu" aria-labelledby="transcript_dropdown">
                                                        {% for annotation in consequences[consequence][gene] %}
                                                            <li role="presentation">
                                                                <a role="menuitem" tabindex="-1" href="/transcript/{{ annotation.Feature }}">
                                                                    {{ annotation.Feature }}
                                                                    {% if annotation.CANONICAL == 'YES' %}
                                                                        *
                                                                    {% endif %}
                                                                    {% if consequence == 'missense_variant' %}
                                                                        ({{ annotation.HGVS }})<br/>
                                                                        &nbsp; &nbsp; &nbsp; Polyphen:
                                                                        {% set polyphen = annotation.PolyPhen.split('(')[0] %}
                                                                        <span class=
                                                                            {% if polyphen == 'benign' %}
                                                                                "bg-success"
                                                                            {% elif polyphen == 'possibly_damaging' %}
                                                                                "bg-warning"
                                                                            {% elif polyphen == 'probably_damaging' %}
                                                                                "bg-danger"
                                                                            {% endif %}>
                                                                        {{ polyphen }}</span>; SIFT:
                                                                        {% set sift = annotation.SIFT.split('(')[0] %}
                                                                        <span class=
                                                                            {% if sift == 'tolerated' %}
                                                                                "bg-success"
                                                                            {% elif sift == 'deleterious' %}
                                                                                "bg-danger"
                                                                            {% endif %}>
                                                                        {{ sift }}</span>
                                                                    {% elif consequence == 'stop_gained' %}
                                                                        ({{ annotation.HGVS }})
                                                                    {% endif %}
                                                                </a>
                                                            </li>
                                                        {% endfor %}
                                                    </ul>
                                                </span>
                                            {% else %}
                                                 - <a href="/transcript/{{ consequences[consequence][gene][0].Feature }}">
                                                    {{ consequences[consequence][gene][0].Feature }}
                                                    {% if consequences[consequence][gene][0].CANONICAL == 'YES' %}
                                                        *
                                                    {% endif %}
                                                </a>
                                                {% if consequence == 'missense_variant' %}
                                                    ({{ consequences[consequence][gene][0].HGVS }})<br/>
                                                    &nbsp; &nbsp; &nbsp; Polyphen:
                                                    {% set polyphen = consequences[consequence][gene][0].PolyPhen.split('(')[0] %}
                                                    <span class=
                                                        {% if polyphen == 'benign' %}
                                                            "bg-success"
                                                        {% elif polyphen == 'possibly_damaging' %}
                                                            "bg-warning"
                                                        {% elif polyphen == 'probably_damaging' %}
                                                            "bg-danger"
                                                        {% endif %}>
                                                    {{ polyphen }}</span>; SIFT:
                                                    {% set sift = consequences[consequence][gene][0].SIFT.split('(')[0] %}
                                                    <span class=
                                                        {% if sift == 'tolerated' %}
                                                            "bg-success"
                                                        {% elif sift == 'deleterious' %}
                                                            "bg-danger"
                                                        {% endif %}>
                                                    {{ sift }}</span>
                                                {% endif %}
                                            {% endif %}
                                        </li>
                                    {% endfor %}
                                </ul>
                            {% endfor %}
                        </div>
                        <small><span class="label label-info">Note:</span> This list may not include additional transcripts in the same gene that the variant does not overlap.</small>
                    {% else %}
                        No annotations were found for this variant.
                    {% endif %}
                </div>
                <div id="browser_container">
                </div>
            </div>

            <div class="col-md-6">
                {% if variant.pop_acs %}
                    <div id="frequency_info_container">
                        <div class="section_header">Population Frequencies</div>
                        <div id="frequency_table_container" class="frequency_displays">
                            {% with chrom = variant.chrom %}
                            <table id="frequency_table">
                                <thead>
                                    <tr>
                                        <th>Population</th>
                                        <th>Allele Count</th>
                                        <th>Allele Number</th>
                                        {% if chrom != 'Y' %}
                                            <th>Number of Homozygotes</th>
                                        {% endif %}
                                        {% if chrom == 'X' or chrom == 'Y' %}
                                            <th>Number of Hemizygotes</th>
                                        {% endif %}
                                        <th>Allele Frequency</th>
                                    </tr>
                                </thead>
                                <tbody>
                                    {% for pop in variant.pop_acs %}
                                        <tr>
                                            <td>{{ pop }}</td>
                                            <td>{{ variant.pop_acs[pop] }}</td>
                                            <td>{{ variant.pop_ans[pop] }}</td>
                                            {% if chrom != 'Y' %}
                                                <td>{{ variant.pop_homs[pop] }}</td>
                                            {% endif %}
                                            {% if chrom == 'X' or chrom == 'Y' %}
                                                <td>{{ variant.pop_hemis[pop] }}</td>
                                            {% endif %}
                                            {% if variant.pop_ans[pop] > 0 %}
                                                <td>{{ '%0.4g' % (variant.pop_acs[pop]/variant.pop_ans[pop]) }}</td>
                                            {% else %}
                                                <td>NA</td>
                                            {% endif %}
                                        </tr>
                                    {% endfor %}
                                </tbody>
                                <tfoot>
                                    <tr>
                                        <td><b>Total</b></td>
                                        <td><b>{{ variant.pop_acs.values()|sum }}</b></td>
                                        <td><b>{{ variant.pop_ans.values()|sum }}</b></td>
                                        {% if chrom != 'Y' %}
                                            <td><b>{{ variant.pop_homs.values()|sum }}</b></td>
                                        {% endif %}
                                        {% if chrom == 'X' or chrom == 'Y' %}
                                            <td><b>{{ variant.pop_hemis.values()|sum }}</b></td>
                                        {% endif %}
                                        <td><b>
                                            {% if variant.allele_freq %}
                                                {{ '%0.4g' % variant.allele_freq }}
                                            {% else %}
                                                NA
                                            {% endif %}
                                        </b></td>
                                    </tr>
                                </tfoot>
                            </table>
                            {% endwith %}
                        </div>
                    </div>
                    <div id="quality_metrics_container">
                        <h3>Quality Metrics</h3>
                        <span class="btn-group" data-toggle="buttons" id="quality_full_site_button_group">
                            <button class="btn btn-primary active quality_full_site_buttons" id="variant_carriers_button"
                                    data-tooltip="Show metric for only individuals with this allele.">
                                <input type="radio"> Variant carriers
                            </button>
                            <button class="btn btn-primary quality_full_site_buttons" id="variant_site_button"
                                    data-tooltip="Show metric for all individuals (at this site, whether or not they have this allele).">
                                <input type="radio" > All individuals
                            </button>
                        </span>
                        <span class="btn-group" data-toggle="buttons" id="quality_display_button_group" style="float: right;">
                            <button class="btn btn-primary active quality_display_buttons" id="genotype_depths_button"
                                    data-tooltip="Per sample depth. Capped at 100X">
                                <input type="radio"> Depth
                            </button>
                            <button class="btn btn-primary quality_display_buttons" id="genotype_qualities_button"
                                    data-tooltip="Per sample genotype quality.">
                                <input type="radio" > Genotype Quality
                            </button>
                        </span>
                        <div id="quality_display_container" class="d3_graph"></div>
                        <small><span class="label label-info">Note:</span> Plot may include low-quality genotypes that were excluded from allele counts in the table above</small>
                        <br/><br/>
                    </div>
                {% else %}
                    {% if any_covered %}
                        <div class="row">
                            <span class="section_header">Coverage</span>
                            {% if base_coverage|length > 1 %}
                                {% include 'coverage_selectors.html' %}
                            {% endif %}
                        </div>
                        <div id="region_coverage"></div>
                    {% else %}
                        This region is not covered in the ExAC dataset.
                    {% endif %}
                {% endif %}
            </div>
        </div>


        {% if read_viz|length > 0  %}
            <div id="igv-container"></div>
	    <br><br>
	    <div>
	      <center>
	      {% if read_viz['het']|length > 1  %}
                  <button class="btn btn-primary quality_full_site_buttons" id="load-more-het-button">load 1 more het sample</button>
	      {% endif %}

   	      {% if read_viz['het']|length > 1 and read_viz['hom']|length > 1 %}  
                  &nbsp;                   &nbsp;                   &nbsp;                   &nbsp; 
              {% endif %}

	      {% if read_viz['hom']|length > 1  %}
                  <button class="btn btn-primary quality_full_site_buttons" id="load-more-hom-button">load 1 more hom sample</button>
	      {% endif %}
              </center>
            </div>
         
	  <!-- IGV dependencies -->

	  <!-- jQuery and jQuery UI -->
	  <script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
	  <link rel="stylesheet" type="text/css" href="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"/> <script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>

	  <!-- Google Fonts and Font awesome -->
	  <link rel="stylesheet" type="text/css" href='//fonts.googleapis.com/css?family=PT+Sans:400,700'> <link rel="stylesheet" type="text/css" href='//fonts.googleapis.com/css?family=Open+Sans'>
	  <link rel="stylesheet" type="text/css" href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css">

	  <!-- Bootstrap JS - for demo only, NOT REQUIRED FOR IGV -->
	  <script src="https://cdnjs.cloudflare.com/ajax/libs/typeahead.js/0.11.1/typeahead.bundle.min.js"></script>
	  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.22.2/js/jquery.tablesorter.js"></script>

	  <!-- igv.js -->
	  <link rel="stylesheet" type="text/css" href="http://igv.org/web/beta/igv-beta.css">
	  <script type="text/javascript" src="http://igv.org/web/beta/igv-beta.min.js"></script>

	  <script type="text/javascript">
	    var read_viz_data = {{ read_viz | tojson | safe }};

            var tracks = [];
	    tracks.push({ url: '/static/read_viz/gencode.v19.sorted.bed', name: "gencode v19" });
	    tracks.push({ url: '/static/read_viz/exome_calling_regions.v1.bed', name: "exac calling intervals" });

	    tracks.push({ url: '/static/read_viz/self_chain.sorted.bed', name: "self chain" });

	    if(read_viz_data["het"] && read_viz_data["het"].length > 0) {
	       tracks.push({ type: 'bam', indexed: true, url: '/static/read_viz/'+read_viz_data["het"][0], name: 'het #1'}); //, height: 200, autoHeight: false});
            }

	    if(read_viz_data["hom"] && read_viz_data["hom"].length > 0) {
	       tracks.push({ type: 'bam', indexed: true, url: '/static/read_viz/'+read_viz_data["hom"][0], name: 'hom #1'}); //, height: 200, autoHeight: false});
            }

	    var tracks_counter = {"het": 1, "hom": 1};
            var load_one_more = function(het_or_hom) {
       	       if(read_viz_data[het_or_hom] && tracks_counter[het_or_hom] < read_viz_data[het_or_hom].length) {
                  var c = tracks_counter[het_or_hom];
                  igv.browser.loadTrack({ type: 'bam', indexed: true, url: '/static/read_viz/'+read_viz_data[het_or_hom][c], name: het_or_hom+' #'+(c+1)});
                  tracks_counter[het_or_hom] += 1;
	          
                  if(tracks_counter[het_or_hom] >= read_viz_data[het_or_hom].length) {
                      $('#load-more-'+het_or_hom+'-button').prop('disabled', true)
                  }
	       }
            }


            var locus = '{{ variant.chrom }}:{{ variant.pos-150 }}-{{ variant.pos+150 }}';
	    var options = {
               showCommandBar: true,
	       genome: 'hg19', 
               locus: locus,
    	       showKaryo: false,
	       tracks: tracks,
            };

            igv.createBrowser(  $("#igv-container")[0],  options);

            //hide the IGV ideogram
	    $(".igv-ideogram-content-div").hide();
	    //$(".igv-logo").hide();
	    
            var trackLabelToggle = $('<div class="igvNavigationToggleTrackLabels">');
            $(".igvNavigationSearch").append("<a id='reset-locus' title='Reset position to "+locus+"'><i class='igv-app-icon fa fa-mail-reply shim-left-6'></i></a>");
         
            // allow one more zoom-in level
            igv.browser.pixelPerBasepairThreshold = function () {
	        return 28.0;  //default is currently 14.0
            };

            //click handlers
	    $('#reset-locus').click(function() {
	       igv.browser.search(locus);
	    });

            $('#load-more-het-button').click(function() {
                load_one_more("het");
	    });
  
            $('#load-more-hom-button').click(function() {		       
                load_one_more("hom");
   	    });

            setInterval(function() {
		igv.browser.trackViews.forEach(function (panel) {
			if(panel.track.name) {
   				     panel.viewportDiv.style.borderBottom = panel.viewportDiv.style.borderLeft = "1px solid #cccccc";
                           if(panel.track.name.match("^h")) {
			       panel.setTrackHeight(300);
                           }
                        }
		});

            }, 1000);
	  </script>
	{% endif %}
    </div>


{% endblock %}
